### PR TITLE
fix: Theme Stylesheet refactor

### LINKED ISSUE #18253: [Task]: App Theming Stylesheet Refactor
### Is there an existing issue for this?

- [X] I have searched the existing issues

### SubTasks

Theme Stylesheets config has become really hard to manage. Need a way to make it simpler and move to the frontend.

### PR DESCRIPTION
**Pull request by jsartisan**
Fixes #18253

Theme Stylesheets config became really hard to manage. 
- Every time a new widget is introduced, we had to update system-themes.json on the backend and get it merged first.
- The system-themes.json file was getting huge and hard to maintain due to the stylesheet. 

This PR moves the theme stylesheets from each to the front end. It moves stylesheet config to individual widgets.

### CODE REVIEW

File: app/client/src/pages/Editor/PropertyPane/PropertyControl.tsx [hunk 1]
```diff
@@ -138,9 +136,8 @@ const PropertyControl = memo((props: Props) => {
    *   theme config and thus it is fetched from there.
    */
   const propertyStylesheetValue = (() => {
-    const widgetStylesheet: AppTheme["stylesheet"][string] = get(
-      selectedTheme,
-      `stylesheet.${widgetProperties.type}`,
+    const widgetStylesheet: AppThemeStylesheet = WidgetFactory.getWidgetStylesheetConfigMap(
```
Comment (ashit-rath) [inline app/client/src/pages/Editor/PropertyPane/PropertyControl.tsx:139]: Do we need to explicitly do `AppThemeStylesheet`? Doesn't `getWidgetStylesheetConfigMap` return type is `AppThemeStylesheet`?

File: app/client/src/sagas/WidgetAdditionSagas.ts [hunk 2]
```diff
@@ -77,21 +70,9 @@ function* getEntityNames() {
  * @returns
  */
 function* getThemeDefaultConfig(type: string) {
```
Comment (ashit-rath) [inline app/client/src/sagas/WidgetAdditionSagas.ts:72]: Do we need this saga? Can't we directly use `WidgetFactory.getWidgetStylesheetConfigMap`?

File: app/client/src/utils/WidgetFactory.tsx [hunk 3]
```diff
@@ -49,6 +50,7 @@ class WidgetFactory {
     readonly PropertyPaneConfig[]
   > = new Map();
   static loadingProperties: Map<WidgetType, Array<RegExp>> = new Map();
+  static stylesheetConfigMap: Map<WidgetType, Record<string, any>> = new Map();
```
Comment (ashit-rath) [inline app/client/src/utils/WidgetFactory.tsx:53]: Should the type be Map<WidgetType, AppThemeStylesheet>?

File: app/client/src/utils/WidgetFactory.tsx [hunk 4]
```diff
@@ -66,6 +68,7 @@ class WidgetFactory {
     propertyPaneStyleConfig?: PropertyPaneConfig[],
     features?: WidgetFeatures,
     loadingProperties?: Array<RegExp>,
+    stylesheetConfig?: Record<string, any>,
```
Comment (ashit-rath) [inline app/client/src/utils/WidgetFactory.tsx:71]: `stylesheetConfig?: AppThemeStylesheet`?

File: app/client/src/utils/WidgetFactory.tsx [hunk 5]
```diff
@@ -263,6 +268,17 @@ class WidgetFactory {
...
     return this.loadingProperties.get(type);
   }
+
+  static getWidgetStylesheetConfigMap(
+    widgetType: WidgetType,
+  ): AppThemeStylesheet {
+    const map = this.stylesheetConfigMap.get(widgetType);
+    if (!map) {
+      log.error("Widget stylesheet properties not defined: ", widgetType);
+      return {};
```
Comment (ashit-rath) [inline app/client/src/utils/WidgetFactory.tsx:278]: Let's return `undefined` from here and let's not explicitly mention the return type of this function as `AppThemeStylesheet`

The consumer of this function would have a better way of handling both the cases.

File: app/client/src/widgets/TableWidgetV2/widget/utilities.ts [hunk 6]
```diff
@@ -513,7 +512,7 @@ export const getSelectedRowBgColor = (accentColor: string) => {
 export const getStylesheetValue = (
   props: TableWidgetProps,
   propertyPath: string,
-  widgetStylesheet?: AppTheme["stylesheet"][string],
+  widgetStylesheet?: Record<string, any>,
```
Comment (ashit-rath) [inline app/client/src/widgets/TableWidgetV2/widget/utilities.ts:515]: widgetStylesheet?: AppThemeStylesheet,

**Review by ashit-rath — COMMENTED** [review]
@jsartisan Can you elaborate the description  of the PR a bit to let folks know the intention of the PR

File: app/client/src/entities/AppTheming/index.ts [hunk 7]
```diff
@@ -1,9 +1,26 @@
...
+    | {
+        [key: string]: string;
+      };
+  childStylesheet?: Stylesheet;
+  resetButtonStyles?: {
+    [key: string]: string;
+  };
+  submitButtonStyles?: {
+    [key: string]: string;
+  };
```
[the remark this answers is not in the document]
Comment (jsartisan) [reply app/client/src/entities/AppTheming/index.ts:21]: Seems like it's not needed. So skipping this one.

File: app/client/src/widgets/TableWidgetV2/widget/utilities.ts (hunk 6 above)
Comment (jsartisan) [reply app/client/src/widgets/TableWidgetV2/widget/utilities.ts]: Done.

File: app/client/src/utils/WidgetFactory.tsx (hunk 5 above)
Comment (jsartisan) [reply app/client/src/utils/WidgetFactory.tsx]: Done.

File: app/client/src/utils/WidgetFactory.tsx (hunk 4 above)
Comment (jsartisan) [reply app/client/src/utils/WidgetFactory.tsx]: Done.

File: app/client/src/utils/WidgetFactory.tsx (hunk 3 above)
Comment (jsartisan) [reply app/client/src/utils/WidgetFactory.tsx]: Done.

File: app/client/src/sagas/WidgetAdditionSagas.ts (hunk 2 above)
Comment (jsartisan) [reply app/client/src/sagas/WidgetAdditionSagas.ts]: Done.

File: app/client/src/pages/Editor/PropertyPane/PropertyControl.tsx (hunk 1 above)
Comment (jsartisan) [reply app/client/src/pages/Editor/PropertyPane/PropertyControl.tsx]: Done.

File: app/client/src/entities/AppTheming/utils.ts [hunk 8]
```diff
@@ -127,6 +130,8 @@ export const getPropertiesToUpdateForReset = (
 
         // reset submit button
         ["submitButtonStyles", "resetButtonStyles"].map((buttonStyleKey) => {
+          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
```
[the remark this answers is not in the document]
Comment (jsartisan) [reply app/client/src/entities/AppTheming/utils.ts]: Done.

**Review by ashit-rath — DISMISSED** [review]

**Comment by chandannkumar** [comment]
@jsartisan 
- Theming is not getting applied on List widget. Please Check the Pagination too
https://bthrujcsw8.vmaker.com/record/yFPShR5Z3ViX9nwg

**Comment by chandannkumar** [comment]
@jsartisan 
- Theming is not getting applied on List widget. Please Check the Pagination too
https://bthrujcsw8.vmaker.com/record/yFPShR5Z3ViX9nwg

**Comment by chandannkumar** [comment]
Tested this PR and LGTM
- Tested for all widgets
- Tested theming related properties for all widgets
- Tested Reset functionalities for theming in relevant widgets
- Tested all widgets with theming on Modals, container type widgets

**Review by dilippitchika — APPROVED** [review]

**Pull request merged**